homeassistant/components/ihc/__init__.py

```diff
@@ -0,0 +1,220 @@
+        vol.All(cv.ensure_list, [
+            vol.All({
+                vol.Required(CONF_XPATH): cv.string,
+                vol.Required(CONF_NODE): cv.string,
+                vol.Optional(CONF_DIMMABLE, default=False): cv.boolean,
+            })
+        ]),
+    vol.Optional(CONF_SENSOR, default=[]):
+        vol.All(cv.ensure_list, [
+            vol.All({
+                vol.Required(CONF_XPATH): cv.string,
+                vol.Required(CONF_NODE): cv.string,
+                vol.Optional(CONF_TYPE, default='Temperature'): cv.string,
+                vol.Optional(CONF_UNIT_OF_MEASUREMENT,
+                             default='°C'): cv.string,
+            })
+        ]),
+    vol.Optional(CONF_SWITCH, default=[]):
+        vol.All(cv.ensure_list, [
+            vol.All({
+                vol.Required(CONF_XPATH): cv.string,
+                vol.Required(CONF_NODE): cv.string,
+            })
+        ]),
+})
+
+SET_RUNTIME_VALUE_BOOL_SCHEMA = vol.Schema({
+    vol.Required(ATTR_IHC_ID): cv.positive_int,
+    vol.Required(ATTR_VALUE): cv.boolean
+})
+
+SET_RUNTIME_VALUE_INT_SCHEMA = vol.Schema({
+    vol.Required(ATTR_IHC_ID): cv.positive_int,
+    vol.Required(ATTR_VALUE): int
+})
+
+SET_RUNTIME_VALUE_FLOAT_SCHEMA = vol.Schema({
+    vol.Required(ATTR_IHC_ID): cv.positive_int,
+    vol.Required(ATTR_VALUE): vol.Coerce(float)
+})
+
+_LOGGER = logging.getLogger(__name__)
+
+IHC_PLATFORMS = ('binary_sensor', 'light', 'sensor', 'switch')
+
+
+def setup(hass, config):
+    """Setup the IHC component."""
+    from ihcsdk.ihccontroller import IHCController
+    conf = config[DOMAIN]
+    url = conf[CONF_URL]
+    username = conf[CONF_USERNAME]
+    password = conf[CONF_PASSWORD]
+    ihc_controller = IHCController(url, username, password)
+
+    if not ihc_controller.authenticate():
+        _LOGGER.error("Unable to authenticate on ihc controller.")
+        return False
+
+    if (conf[CONF_AUTOSETUP] and
+            not autosetup_ihc_products(hass, ihc_controller)):
+        return False
+
+    hass.data[IHC_DATA] = {
+        IHC_CONTROLLER: ihc_controller,
+        IHC_INFO: conf[CONF_INFO]}
+
+    setup_service_functions(hass, ihc_controller)
+    return True
+
+
+def autosetup_ihc_products(hass: HomeAssistantType, ihc_controller):
+    """Auto setup of IHC products from the ihc project file."""
+    project_xml = ihc_controller.get_project()
+    if not project_xml:
+        _LOGGER.error("Unable to read project from ihc controller.")
+        return False
+    project = xml.etree.ElementTree.fromstring(project_xml)
+
+    # if an auto setup file exist in the configuration it will override
+    yaml_path = hass.config.path(AUTO_SETUP_YAML)
+    if not os.path.isfile(yaml_path):
+        yaml_path = os.path.join(os.path.dirname(__file__), AUTO_SETUP_YAML)
+    yaml = load_yaml_config_file(yaml_path)
+    try:
+        auto_setup_conf = AUTO_SETUP_SCHEMA(yaml)
+    except VoluptuousError as exception:
```

Comment:
The except VoluptuousError will catch all error from voluptuous (I copied it from somewhere else in home-assistant), but I assume your reason is that vol.Invalid is better because the schema is always valid (build into the code). I have changed it to vol.Invalid